src/MediaWiki/MessageBuilder.php

```diff
@@ -98,11 +100,31 @@ public function listToCommaSeparatedText( array $list ) {
 	 * @return string
 	 */
 	public function prevNextToText( Title $title, $limit, $offset, array $query, $isAtTheEnd ) {
-		if ( class_exists( 'MediaWiki\Navigation\PrevNextNavigationRenderer' ) ) {
-			$prevNext = new PrevNextNavigationRenderer( \RequestContext::getMain() );
-			return $prevNext->buildPrevNextNavigation( $title, $offset, $limit, $query, $isAtTheEnd );
+		if ( class_exists( PagerNavigationBuilder::class ) ) {
+			// MW > 1.39
+			$navBuilder = new PagerNavigationBuilder( RequestContext::getMain() );
+			$navBuilder
+				->setPage( $title )
+				->setLinkQuery( [ 'limit' => $limit, 'offset' => $offset ] + $query )
+				->setLimitLinkQueryParam( 'limit' )
+				->setCurrentLimit( $limit )
+				->setPrevTooltipMsg( 'prevn-title' )
+				->setNextTooltipMsg( 'nextn-title' )
+				->setLimitTooltipMsg( 'shown-title' );
+
+			if ( $offset > 0 ) {
+				$navBuilder->setPrevLinkQuery( [ 'offset' => (string)max( $offset - $limit, 0 ) ] );
+			}
+
+			if ( !$isAtTheEnd ) {
+				$navBuilder->setNextLinkQuery( [ 'offset' => (string)( $offset + $limit ) ] );
+			}
+
+			return $navBuilder->getHtml();
```

Comment:
I think it is much cleaner if you put the body of the if statement into its own private method.

Or, alternatively, perhaps the if-else is not needed, and the method body can be replaced by the if body. Would need to check that this works down to MW 35 though, so is more work, but ends up with simpler code.

Reply:
I still recommend you extract into a private method.